mlir/include/mlir/Dialect/LLVMIR/GENXOps.td

```diff
@@ -741,4 +741,43 @@ def GENX_MatrixMapOp : GENX_Op<"matrix.map", [
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// Assert Op
+//===----------------------------------------------------------------------===//
+
+def GENX_AssertOp : GENX_Op<"assert">, 
+  Arguments<(ins LLVM_i8Ptr_generic: $msg_ptr,
+             LLVM_i8Ptr_generic: $file_ptr,
+             I32: $line_number,
+             LLVM_i8Ptr_generic: $func_ptr)> {
+  let summary = "Workitem assert";
+
+  string baseDescription = [{
+    The `genx.assert` operation performs a workitem assert message with file
+    name, line number and function name for its location
+  }];
+
+  string llvmBuilder = [{
+    llvm::Type *retType = builder.getVoidTy();
+    llvm::Type *ptrType = builder.getInt8PtrTy(/* spirv::Generic */4);
+    llvm::Type *lnType = builder.getInt32Ty();
+    llvm::CallInst *ci = createDeviceFunctionCall(
+                            builder, "__assert_fail", retType,
+                            {ptrType, ptrType, lnType, ptrType}, 
+                            {$msg_ptr, $file_ptr, $line_number, $func_ptr});
+    ci->addFnAttr(llvm::Attribute::Convergent);
```

Comment:
Would it work if we set the function attributes in LLVM MLIR dialect like below?
```
attributes {llvm.cconv = #llvm.cconv<spir_funccc>, passthrough = ["convergent", "nounwind"]}
```

Reply:
The syntax is not accepted.  error: stray ‘#’ in program
      353 |       ret.setAttrs({llvm.cconv = #llvm.cconv<spir_funccc>, passthrough = ["convergent", "nounwind"]});

Reply:
It would be great if we can set the function attributes directly in LLVM MLIR dialect though.  I just don't know the correct syntax for the attributes and the interface functions.

Reply:
In the SYCL-MLIR project, we created a AttrBuilder (https://github.com/intel/llvm/blob/sycl-mlir/polygeist/tools/cgeist/Lib/Attributes.cc) class to add attributes more easily. 
For `attributes {llvm.cconv = #llvm.cconv<spir_funccc>, passthrough = ["convergent", "nounwind"]}`, there are 2 attributes, one as `llvm.cconv` and one as `passthrough`. 
SYCL-MLIR use case is to always add these attributes for a function, i.e. the callee function in our case here. I wonder if just adding these attributes for the callee function is enough, or we need the attributes in both the call instruction and the callee function.

Reply:
I think the syntax without `AttrBuilder` is like:
```
  callee->setAttr("passthrough", builder.getArrayAttr(
                                   StringAttr::get(ctx, "convergent")));
```

Reply:
forwardPassthroughAttributes https://github.com/intel/llvm/blob/sycl-mlir/mlir/lib/Target/LLVMIR/ModuleTranslation.cpp#L830 in module translation is going to forward the passthrough attributes from MLIR to LLVMIR.

Reply:
I am able to set the SPIR_FUNC calling convention to the FuncOp using func.setCConv(LLVM::cconv::CConv::SPIR_FUNC). However, without setting the CallOp to SPIR_FUNC calling convention setting, the call statement is still removed by LLVM optimizer.  I will just go with the customized GENX AssertOp for now and visiting this later.

Reply:
Created a PR to cherry-pick the commit that add cconv to CallOp: https://github.com/pengtu/llvm-project/pull/45
Please let me know if that's preferable.